crates/xline/src/storage/db.rs

```diff
@@ -56,76 +59,46 @@ impl DB {
         }))
     }
+}
 
-    /// Get del lease key buffer
+impl StorageOps for DB {
     #[inline]
-    fn get_del_lease_key_buffer(ops: &[WriteOp]) -> HashMap<i64, Vec<u8>> {
-        ops.iter()
-            .filter_map(|op| {
-                if let WriteOp::DeleteLease(lease_id) = *op {
-                    Some((lease_id, lease_id.encode_to_vec()))
-                } else {
-                    None
-                }
-            })
-            .collect::<HashMap<_, _>>()
+    fn write(&self, op: WriteOperation<'_>, sync: bool) -> Result<(), engine::EngineError> {
+        self.engine.write(op, sync)
     }
 
-    /// get del alarm buffer
     #[inline]
-    fn get_del_alarm_buffer(ops: &[WriteOp]) -> Vec<u8> {
-        ops.iter()
-            .find_map(|op| {
-                if let WriteOp::DeleteAlarm(ref key) = *op {
-                    Some(key.encode_to_vec())
-                } else {
-                    None
-                }
-            })
-            .unwrap_or_default()
-    }
-
-    /// Get values by keys from storage
-    ///
-    /// # Errors
-    ///
-    /// if error occurs in storage, return `Err(error)`
-    pub(crate) fn get_values<K>(
-        &self,
-        table: &'static str,
-        keys: &[K],
-    ) -> Result<Vec<Option<Vec<u8>>>, ExecuteError>
+    fn write_multi<'a, Ops>(&self, ops: Ops, sync: bool) -> Result<(), engine::EngineError>
     where
-        K: AsRef<[u8]> + std::fmt::Debug + Sized,
+        Ops: IntoIterator<Item = WriteOperation<'a>>,
     {
-        let values = self
-            .engine
-            .get_multi(table, keys)
-            .map_err(|e| ExecuteError::DbError(format!("Failed to get keys {keys:?}: {e}")))?
-            .into_iter()
-            .collect::<Vec<_>>();
-
-        assert_eq!(values.len(), keys.len(), "Index doesn't match with DB");
+        self.engine.write_multi(ops, sync)
+    }
 
-        Ok(values)
+    #[inline]
+    fn get(
+        &self,
+        table: &str,
+        key: impl AsRef<[u8]>,
+    ) -> Result<Option<Vec<u8>>, engine::EngineError> {
+        self.engine.get(table, key)
     }
 
-    /// Get values by keys from storage
-    ///
-    /// # Errors
-    ///
-    /// if error occurs in storage, return `Err(error)`
-    pub(crate) fn get_value<K>(
+    #[inline]
+    fn get_multi(
         &self,
-        table: &'static str,
-        key: K,
-    ) -> Result<Option<Vec<u8>>, ExecuteError>
-    where
-        K: AsRef<[u8]> + std::fmt::Debug,
-    {
-        self.engine
-            .get(table, key.as_ref())
-            .map_err(|e| ExecuteError::DbError(format!("Failed to get key {key:?}: {e}")))
+        table: &str,
+        keys: &[impl AsRef<[u8]>],
+    ) -> Result<Vec<Option<Vec<u8>>>, engine::EngineError> {
+        self.engine.get_multi(table, keys)
+    }
+}
+
+impl DB {
+    /// Creates a transaction
+    #[allow(unused)] // TODO: use this in the following refactor
+    pub(crate) fn transaction(&self) -> Transaction {
```

Comment:
Is this a local transaction provided by rocksdb? How will we use it (where could we use it) in the following refactor ?

Reply:
Yes, it will be used in speculative execution to provide a temporary execution environment. In speculative execution, the results are not committed to the state machine, so we need to discard the transaction once we obtain the result.